src/io/hdf5_read_utils.c

```diff
@@ -0,0 +1,161 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "hdf5_read_utils.h"
+
+herr_t read_attribute(hid_t fd, const char *group_name, const char *attr_name, void *attribute, const size_t dst_size)
+{
+    hid_t attr_id = H5Aopen_by_name(fd, group_name, attr_name, H5P_DEFAULT, H5P_DEFAULT);
+    if (attr_id < 0) {
+        fprintf(stderr, "Error:Could not open the attribute '%s' in group '%s'\n", attr_name, group_name);
+        return attr_id;
+    }
+
+    hid_t attr_dtype = H5Aget_type(attr_id);
+    if(attr_dtype < 0) {
+        fprintf(stderr, "Could not get the datatype for the attribute '%s' in group '%s'\n", attr_name, group_name);
+        return attr_dtype;
+    }
+
+    if(dst_size != H5Tget_size(attr_dtype)) {
+        fprintf(stderr,"Error while reading %s attribute in group %s\n", group_name, attr_name);
+        fprintf(stderr,"The HDF5 attribute has with size %zu bytes into destination with size = %zu\n",
+                H5Tget_size(attr_dtype), dst_size);
+        fprintf(stderr,"Perhaps the size of the destination datatype needs to be updated?\n");
+        return -1;
+    }
+
+    herr_t status = H5Aread(attr_id, attr_dtype, attribute);
+    if (status < 0) {
+        fprintf(stderr, "Could not read the attribute %s in group %s\n", attr_name, group_name);
+        H5Eprint(attr_id, stderr);
+        return status;
+    }
+
+    status = H5Tclose(attr_dtype);
+    if (status < 0) {
+        fprintf(stderr, "Error when closing the datatype for the attribute '%s' in group '%s'.\n", attr_name, group_name);
+        H5Eprint(status, stderr);
+        return status;
+    }
+
+    status = H5Aclose(attr_id);
+    if (status < 0) {
+        fprintf(stderr, "Error when closing the attribute %s in group %s.\n", attr_name, group_name);
+        H5Eprint(attr_id, stderr);
+        return status;
+    }
+
+    return (herr_t) EXIT_SUCCESS;
+}
+
+
+herr_t read_dataset_shape(hid_t fd, const char *dataset_name, int *ndims, hsize_t **dims)
+{
+    hid_t dataset_id = H5Dopen(fd, dataset_name, H5P_DEFAULT);
+    if(dataset_id < 0) {
+        fprintf(stderr, "Error encountered when trying to open up dataset '%s'.\n", dataset_name);
+        H5Eprint(dataset_id, stderr);
+        return -1;
+    }
+
+    hid_t dspace = H5Dget_space(dataset_id);
+    if(dspace < 0) {
+        fprintf(stderr, "Error encountered when trying to get dataspace for dataset '%s'.\n", dataset_name);
+        H5Eprint(dataset_id, stderr);
+        return dspace;
+    }
+
+    *ndims = H5Sget_simple_extent_ndims(dspace);
+    if(*ndims < 0) {
+        fprintf(stderr,"Error: Could not get the number of dimensions of the dataset '%s'\n", dataset_name);
+        return (herr_t) EXIT_FAILURE;
+    }
+
+    *dims = calloc(*ndims, sizeof(**dims));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Either the condition 'dims==NULL' is redundant or there is possible null pointer dereference: dims.](https://app.codacy.com/app/jacobseiler/sage-model/pullRequest?prid=4620074)